File: admin-web/src/api/skills.ts
import http from './http'

export type SkillLevel = 'L1' | 'L2' | 'L3'

export interface Skill {
  id: number
  name: string             // 认证名称 e.g. CCIE 路由交换
  category: string         // 分类 e.g. 网络能力
  issuer?: string | null   // 厂商 e.g. Cisco
  level?: SkillLevel | null
  is_active: boolean
}

export type SkillPayload = Omit<Skill, 'id'>

export interface SkillBulkItem {
  issuer: string
  name: string
  level: SkillLevel
}

export interface SkillBulkImportPayload {
  category: string
  items: SkillBulkItem[]
}

export interface SkillBulkResult {
  created: number
  skipped: number
  skipped_names: string[]
}

export const listSkills = () => http.get<Skill[]>('/skills').then((r) => r.data)
export const createSkill = (p: Partial<SkillPayload>) =>
  http.post<Skill>('/skills', p).then((r) => r.data)
export const updateSkill = (id: number, p: Partial<SkillPayload>) =>
  http.patch<Skill>(`/skills/${id}`, p).then((r) => r.data)
export const deleteSkill = (id: number) => http.delete(`/skills/${id}`)
export const bulkImportSkills = (p: SkillBulkImportPayload) =>
  http.post<SkillBulkResult>('/skills/bulk-import', p).then((r) => r.data)

File: backend/app/api/admin/skills.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.deps import get_current_user, require_role
from app.models.skill import Skill
from app.schemas.skill import (
    SkillBulkImport, SkillBulkResult,
    SkillCreate, SkillOut, SkillUpdate,
)

router = APIRouter(prefix="/skills", tags=["skills"])


@router.get("", response_model=list[SkillOut])
async def list_skills(
    db: AsyncSession = Depends(get_db),
    _: dict = Depends(get_current_user),
) -> list[SkillOut]:
    result = await db.execute(select(Skill).order_by(Skill.category, Skill.name))
    return [SkillOut.model_validate(s) for s in result.scalars().all()]


@router.post("", response_model=SkillOut, status_code=status.HTTP_201_CREATED)
async def create_skill(
    payload: SkillCreate,
    db: AsyncSession = Depends(get_db),
    _: dict = Depends(require_role("admin", "lead", "pm")),
) -> SkillOut:
    if (await db.execute(select(Skill).where(Skill.name == payload.name))).scalar_one_or_none():
        raise HTTPException(status_code=400, detail="技能已存在")
    s = Skill(**payload.model_dump())
    db.add(s)
    await db.commit()
    await db.refresh(s)
    return SkillOut.model_validate(s)


@router.patch("/{skill_id}", response_model=SkillOut)
async def update_skill(
    skill_id: int,
    payload: SkillUpdate,
    db: AsyncSession = Depends(get_db),
    _: dict = Depends(require_role("admin", "lead", "pm")),
) -> SkillOut:
    s = await db.get(Skill, skill_id)
    if not s:
        raise HTTPException(status_code=404, detail="技能不存在")
    for k, val in payload.model_dump(exclude_unset=True).items():
        setattr(s, k, val)
    await db.commit()
    await db.refresh(s)
    return SkillOut.model_validate(s)


@router.post("/bulk-import", response_model=SkillBulkResult)
async def bulk_import_skills(
    payload: SkillBulkImport,
    db: AsyncSession = Depends(get_db),
    _: dict = Depends(require_role("admin", "lead", "pm")),
) -> SkillBulkResult:
    """批量导入认证目录：整批共用 category，每行含 issuer + name + level。
    遇到重名（name 已存在）跳过，不报错。"""
    existing = {
        s.name for s in (await db.execute(select(Skill.name))).scalars().all()
    } if False else set(
        (await db.execute(select(Skill.name))).scalars().all()
    )
    created = 0
    skipped: list[str] = []
    for it in payload.items:
        if it.name in existing:
            skipped.append(it.name)
            continue
        s = Skill(
            name=it.name, category=payload.category,
            issuer=it.issuer, level=it.level, is_active=True,
        )
        db.add(s)
        existing.add(it.name)
        created += 1
    await db.commit()
    return SkillBulkResult(created=created, skipped=len(skipped), skipped_names=skipped)


@router.delete("/{skill_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_skill(
    skill_id: int,
    db: AsyncSession = Depends(get_db),
    _: dict = Depends(require_role("admin", "lead")),
) -> None:
    s = await db.get(Skill, skill_id)
    if not s:
        raise HTTPException(status_code=404, detail="技能不存在")
    await db.delete(s)
    await db.commit()
